jsonrpc/ipc.go

```diff
@@ -0,0 +1,178 @@
+package jsonrpc
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/p2p/netutil"
+	"github.com/sourcegraph/conc"
+)
+
+type Ipc struct {
+	rpc    *Server
+	events NewRequestListener
+	log    utils.SimpleLogger
+
+	connWg sync.WaitGroup // connWg is a WaitGroup for tracking active connections.
+
+	connParams IpcConnParams
+	listener   net.Listener
+
+	// everything below is protected
+	mu    sync.Mutex
+	conns map[net.Conn]struct{} // conns is a map that holds active connections.
+}
+
+// NewIpc creates a new IPC handler instance with the provided RPC server, endpoint and logger.
+func NewIpc(rpc *Server, endpoint string, log utils.SimpleLogger) (*Ipc, error) {
+	listener, err := createListener(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	return &Ipc{
+		rpc:        rpc,
+		log:        log,
+		connParams: DefaultIpcConnParams(),
+		conns:      make(map[net.Conn]struct{}),
+		events:     &SelectiveListener{},
+		listener:   listener,
+	}, nil
+}
+
+// WithConnParams sanity checks and applies the provided params.
+func (i *Ipc) WithConnParams(p *IpcConnParams) *Ipc {
+	i.connParams = *p
+	return i
+}
+
+// WithListener registers a NewRequestListener
+func (i *Ipc) WithListener(listener NewRequestListener) *Ipc {
+	i.events = listener
+	return i
+}
+
+// Run launches the IPC handler and handles any potential errors.
+// It is the caller's responsibility to cancel the provided context when they wish to
+// gracefully shut down the IPC handler.
+func (i *Ipc) Run(ctx context.Context) error {
+	var wg conc.WaitGroup
+	defer wg.Wait()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	wg.Go(func() {
+		defer cancel()
+		for {
+			conn, err := i.listener.Accept()
+			if netutil.IsTemporaryError(err) {
+				i.log.Warnw("Failed to accept connection", "err", err)
+				continue
+			} else if err != nil {
+				i.log.Warnw("Accept connection", "err", err)
+				return
+			}
```

Comment:
This warning gets logged on a simple ctrl-c. Is that avoidable?
```
21:58:13.023 14/12/2023 -06:00	WARN	jsonrpc/ipc.go:78	Accept connection	{"err": "accept unix /home/user/repos/juno/juno/juno_v0_5.ipc: use of closed network connection"}
21:58:13.023 14/12/2023 -06:00	WARN	jsonrpc/ipc.go:78	Accept connection	{"err": "accept unix /home/user/repos/juno/juno/juno_v0_4.ipc: use of closed network connection"}
21:58:13.023 14/12/2023 -06:00	WARN	jsonrpc/ipc.go:78	Accept connection	{"err": "accept unix /home/user/repos/juno/juno/juno.ipc: use of closed network connection"}
```

Reply:
Preferably we avoid this somehow. If it's too difficult, we can demote it to a debug log.

Reply:
Sure, go ahead for the adjustment/rebase. Regarding the changes I'll prolly take a look at it later today.